source/dhtslib/vcf/header.d

```diff
@@ -61,4 +318,415 @@ struct VCFHeader
+    
+    /** Add INFO (§1.2.2) or FORMAT (§1.2.4) tag
+
+    The INFO tag describes row-specific keys used in the INFO column;
+    The FORMAT tag describes sample-specific keys used in the last, and optional, genotype column.
+
+    Template parameter: string; must be INFO or FORMAT
+
+    The first four parameters are required; NUMBER and TYPE have specific allowable values.
+    source and version are optional, but recommended (for INFO only).
+
+    *   id:     ID tag
+    *   number: NUMBER tag; here a string because it can also take special values {A,R,G,.} (see §1.2.2)
+    *   type:   Integer, Float, Flag, Character, and String
+    *   description: Text description; will be double quoted
+    *   source:      Annotation source  (eg dbSNP)
+    *   version:     Annotation version (eg 142)
+    */
+
+    void addHeaderLine(HeaderRecordType lineType, T)(string id, T number, HeaderTypes type,
+                                    string description="",
+                                    string source="",
+                                    string _version="")
+    if((isIntegral!T || is(T == HeaderLengths)) && lineType != HeaderRecordType.NULL )       
+    {
+        HeaderRecord rec;
+        rec.setHeaderRecordType = lineType;
+        rec.setID(id);
+        rec.setLength(number);
+        rec.setValueType(type);
+        static if(lineType == HeaderRecordType.INFO || lineType == HeaderRecordType.FILTER || lineType == HeaderRecordType.FORMAT){
+            if(description == ""){
+                throw new Exception("description cannot be empty for " ~ HeaderRecordTypeStrings[lineType]);    
+            }
+        }
+        rec.setDescription(description);
+        if(source != "")
+            rec["source"] = "\"%s\"".format(source);
+        if(_version != "")
+            rec["version"] = "\"%s\"".format(_version);
+
+        this.addHeaderRecord(rec);
+    }
+
+    /** Add FILTER tag (§1.2.3) */
+    void addHeaderLine(HeaderRecordType lineType)(string id, string description)
+    if(lineType == HeaderRecordType.FILTER)
+    {
+        HeaderRecord rec;
+        rec.setHeaderRecordType = lineType;
+        rec.setID(id);
+        rec.setDescription("\"%s\"".format(description));
+
+        this.addHeaderRecord(rec);
+    }
+
+    /** Add FILTER tag (§1.2.3) */
+    deprecated void addFilter(string id, string description)
+    {
+        addHeaderLine!(HeaderRecordType.FILTER)(id, description);
+    }
+
+    /// string representation of header
+    string toString(){
+        import htslib.kstring;
+        kstring_t s;
+
+        const int ret = bcf_hdr_format(this.hdr, 0, &s);
+        if (ret)
+        {
+            hts_log_error(__FUNCTION__,
+                format("bcf_hdr_format returned nonzero (%d) (likely EINVAL, invalid bcf_hdr_t struct?)", ret));
+            return "[VCFHeader bcf_hdr_format parse_error]";
+        }
+
+        return cast(string) s.s[0 .. s.l];
+    }
+}
+
+///
+debug(dhtslib_unittest)
+unittest

```

Comment:
Since no VCF file is actually used, this is a great unit test

Reply:
Yeah I need to go through the library and convert many of our "integration" tests that rely on the htslib test files to true unit tests.

Reply:
Will leave for another day